File: server/job/src/main/java/io/iamcyw/ams/job/predicate/JobPredicateService.java
package io.iamcyw.ams.job.predicate;

import io.iamcyw.ams.domain.AlarmMessage;
import io.iamcyw.ams.domain.job.predicate.usecase.QueryAllowStrategy;
import io.iamcyw.ams.domain.job.strategy.entity.StrategyPredicate;
import io.iamcyw.ams.job.strategy.repository.StrategyPredicateRepository;
import io.iamcyw.tower.messaging.QueryHandle;
import io.iamcyw.tower.messaging.UseCase;
import io.iamcyw.tower.messaging.gateway.MessageGateway;
import org.mvel2.MVEL;

import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

@UseCase
public class JobPredicateService {

    protected final MessageGateway messageGateway;

    protected final StrategyPredicateRepository predicateRepository;

    public JobPredicateService(MessageGateway messageGateway, StrategyPredicateRepository predicateRepository) {
        this.messageGateway = messageGateway;
        this.predicateRepository = predicateRepository;
    }

    @QueryHandle
    public Set<Long> query(QueryAllowStrategy queryAllowStrategy) {
        List<StrategyPredicate> policyList = predicateRepository.queryBySource(queryAllowStrategy.source());

        return policyList.stream().parallel().filter(policy -> predicatePayLoad(queryAllowStrategy.payload(), policy))
                         .map(strategy -> strategy.getStrategy().getId()).collect(Collectors.toUnmodifiableSet());
    }

    private boolean predicatePayLoad(AlarmMessage message, StrategyPredicate predicate) {
        return switch (predicate.getType()) {
            case EXPRESS -> express(predicate.getExpress(), message);
            case PASS -> true;
        };
    }

    private boolean express(String express, AlarmMessage alarmMessage) {
        return (boolean) MVEL.eval(express, alarmMessage);
    }

}

File: server/job/src/main/kotlin/io/iamcyw/ams/job/predicate/JobPredicateService.kt
package io.iamcyw.ams.job.predicate

import io.iamcyw.ams.domain.job.express
import io.iamcyw.ams.domain.job.fixed
import io.iamcyw.ams.domain.job.predicate.PredicatePayLoad
import io.iamcyw.ams.domain.job.predicate.QueryAllowStrategy
import io.iamcyw.ams.domain.job.strategy.PolicyContrastType.*
import io.iamcyw.ams.domain.job.strategy.StrategyPolicyPO
import io.iamcyw.tower.commandhandling.gateway.CommandGateway
import io.iamcyw.tower.queryhandling.QueryHandle
import io.iamcyw.tower.queryhandling.gateway.QueryGateway
import javax.enterprise.context.ApplicationScoped
import javax.inject.Inject

@ApplicationScoped
class JobPredicateService @Inject constructor(
    private val commandGateway: CommandGateway,
    private val queryGateway: QueryGateway
) {

    @QueryHandle
    fun query(queryAllowStrategy: QueryAllowStrategy): List<Long> {
        val policies = StrategyPolicyPO.queryBySource(queryAllowStrategy.source)
        return policies.filter {
            val predicatePayLoad = PredicatePayLoad(queryAllowStrategy.payload, it.id!!)
            queryGateway.query(predicatePayLoad)
        }.flatMap {
            it.strategy.map { strategy ->
                strategy.id!!
            }
        }.distinct()
    }

    @QueryHandle
    fun query(predicatePayLoad: PredicatePayLoad): Boolean {
        val policyPO = StrategyPolicyPO.findById(predicatePayLoad.policy)
            ?: throw IllegalStateException("the alarm(${predicatePayLoad.policy}) no found")

        return policyPO.metadata.all {

            when (it.contrastType) {
                Pass -> {
                    true
                }
                Fixed -> predicatePayLoad.message.fixed(it.scope, it.contrastKey, it.contrastValue)
                Express -> predicatePayLoad.message.express(
                    it.scope,
                    it.contrastKey,
                    it.contrastValue
                )
                null -> false
            }
        }
    }

}